From 645ed48af9e2a804e04246d25c848961cd058aac Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Wed, 21 Dec 2016 14:35:15 +0200
Subject: [PATCH 03/11] Revert "Bug 1114594 - Remove promptForSaveToFile in
 favor of promptForSaveToFileAsync. r=paolo"

This reverts commit f5fe5678d69887fd2ffce1729028d0a2e676be2f.

Signed-off-by: Raine Makelainen <raine.makelainen@jolla.com>
---
 b2g/components/HelperAppDialog.js                  |  8 +++++++
 mobile/android/components/HelperAppDialog.js       |  5 ++++
 .../downloads/test/unit/downloads_manifest.js      |  4 +++-
 toolkit/components/jsdownloads/test/unit/head.js   |  9 +++++++
 toolkit/mozapps/downloads/nsHelperAppDlg.js        | 16 +++++++++++++
 .../exthandler/nsExternalHelperAppService.cpp      | 24 ++++++++++++-------
 .../exthandler/nsIHelperAppLauncherDialog.idl      | 28 +++++++++++++++++++++-
 .../tests/mochitest/test_unsafeBidiChars.xhtml     |  3 +++
 .../exthandler/tests/unit_ipc/test_encoding.js     |  4 +++-
 9 files changed, 90 insertions(+), 11 deletions(-)

diff --git a/b2g/components/HelperAppDialog.js b/b2g/components/HelperAppDialog.js
index a155728..71d5423 100644
--- a/b2g/components/HelperAppDialog.js
+++ b/b2g/components/HelperAppDialog.js
@@ -33,6 +33,14 @@ HelperAppLauncherDialog.prototype = {
     aLauncher.saveToDisk(null, false);
   },
 
+  promptForSaveToFile: function(aLauncher,
+                                aContext,
+                                aDefaultFile,
+                                aSuggestedFileExt,
+                                aForcePrompt) {
+    throw Cr.NS_ERROR_NOT_AVAILABLE;
+  },
+
   promptForSaveToFileAsync: function(aLauncher,
                                      aContext,
                                      aDefaultFile,
diff --git a/mobile/android/components/HelperAppDialog.js b/mobile/android/components/HelperAppDialog.js
index 8706d73..f4047c4 100644
--- a/mobile/android/components/HelperAppDialog.js
+++ b/mobile/android/components/HelperAppDialog.js
@@ -245,6 +245,11 @@ HelperAppLauncherDialog.prototype = {
       Services.prefs.clearUserPref(this._getPrefName(mime));
   },
 
+  promptForSaveToFile: function () {
+    throw new Components.Exception("Async version must be used",
+                                   Cr.NS_ERROR_NOT_AVAILABLE);
+  },
+
   promptForSaveToFileAsync: function (aLauncher, aContext, aDefaultFile,
                                       aSuggestedFileExt, aForcePrompt) {
     Task.spawn(function* () {
diff --git a/toolkit/components/downloads/test/unit/downloads_manifest.js b/toolkit/components/downloads/test/unit/downloads_manifest.js
index 7bef411..a24622a 100644
--- a/toolkit/components/downloads/test/unit/downloads_manifest.js
+++ b/toolkit/components/downloads/test/unit/downloads_manifest.js
@@ -15,7 +15,9 @@ HelperAppDlg.prototype = {
   show: function (launcher, ctx, reason) {
     launcher.MIMEInfo.preferredAction = Ci.nsIMIMEInfo.saveToDisk;
     launcher.launchWithApplication(null, false);
-  }
+  },
+
+  promptForSaveToFile: function (launcher, ctx, defaultFile, suggestedExtension, forcePrompt) { }
 }
 
 
diff --git a/toolkit/components/jsdownloads/test/unit/head.js b/toolkit/components/jsdownloads/test/unit/head.js
index 64a330d..bfed5c7 100644
--- a/toolkit/components/jsdownloads/test/unit/head.js
+++ b/toolkit/components/jsdownloads/test/unit/head.js
@@ -804,6 +804,15 @@ add_task(function test_common_initialize()
   // saved to disk without asking for a destination interactively.
   let mock = {
     QueryInterface: XPCOMUtils.generateQI([Ci.nsIHelperAppLauncherDialog]),
+    promptForSaveToFile: function (aLauncher, aWindowContext,
+                                   aDefaultFileName,
+                                   aSuggestedFileExtension,
+                                   aForcePrompt)
+    {
+      throw new Components.Exception(
+                         "Synchronous promptForSaveToFile not implemented.",
+                         Cr.NS_ERROR_NOT_AVAILABLE);
+    },
     promptForSaveToFileAsync(aLauncher,
                              aWindowContext,
                              aDefaultFileName,
diff --git a/toolkit/mozapps/downloads/nsHelperAppDlg.js b/toolkit/mozapps/downloads/nsHelperAppDlg.js
index 6e847f3..eb6c844 100644
--- a/toolkit/mozapps/downloads/nsHelperAppDlg.js
+++ b/toolkit/mozapps/downloads/nsHelperAppDlg.js
@@ -204,6 +204,22 @@ nsUnknownContentTypeDialog.prototype = {
                    bundle.GetStringFromName("badPermissions"));
   },
 
+  // promptForSaveToFile:  Display file picker dialog and return selected file.
+  //                       This is called by the External Helper App Service
+  //                       after the ucth dialog calls |saveToDisk| with a null
+  //                       target filename (no target, therefore user must pick).
+  //
+  //                       Alternatively, if the user has selected to have all
+  //                       files download to a specific location, return that
+  //                       location and don't ask via the dialog.
+  //
+  // Note - this function is called without a dialog, so it cannot access any part
+  // of the dialog XUL as other functions on this object do.
+
+  promptForSaveToFile: function(aLauncher, aContext, aDefaultFile, aSuggestedFileExtension, aForcePrompt) {
+    throw new Components.Exception("Async version must be used", Components.results.NS_ERROR_NOT_AVAILABLE);
+  },
+
   promptForSaveToFileAsync: function(aLauncher, aContext, aDefaultFile, aSuggestedFileExtension, aForcePrompt) {
     var result = null;
 
diff --git a/uriloader/exthandler/nsExternalHelperAppService.cpp b/uriloader/exthandler/nsExternalHelperAppService.cpp
index fc2f2385..39f066a 100644
--- a/uriloader/exthandler/nsExternalHelperAppService.cpp
+++ b/uriloader/exthandler/nsExternalHelperAppService.cpp
@@ -2300,16 +2300,24 @@ void nsExternalAppHandler::RequestSaveDestination(const nsAFlatString &aDefaultF
   // picker is up would cause Cancel() to be called, and the dialog would be
   // released, which would release this object too, which would crash.
   // See Bug 249143
+  nsIFile* fileToUse;
   RefPtr<nsExternalAppHandler> kungFuDeathGrip(this);
   nsCOMPtr<nsIHelperAppLauncherDialog> dlg(mDialog);
-
-  rv = dlg->PromptForSaveToFileAsync(this,
-                                     GetDialogParent(),
-                                     aDefaultFile.get(),
-                                     aFileExtension.get(),
-                                     mForceSave);
-  if (NS_FAILED(rv)) {
-    Cancel(NS_BINDING_ABORTED);
+  rv = dlg->PromptForSaveToFile(this,
+                                GetDialogParent(),
+                                aDefaultFile.get(),
+                                aFileExtension.get(),
+                                mForceSave, &fileToUse);
+
+  if (rv == NS_ERROR_NOT_AVAILABLE) {
+    // we need to use the async version -> nsIHelperAppLauncherDialog.promptForSaveToFileAsync.
+    rv = mDialog->PromptForSaveToFileAsync(this, 
+                                           GetDialogParent(),
+                                           aDefaultFile.get(),
+                                           aFileExtension.get(),
+                                           mForceSave);
+  } else {
+    SaveDestinationAvailable(rv == NS_OK ? fileToUse : nullptr);
   }
 }
 
diff --git a/uriloader/exthandler/nsIHelperAppLauncherDialog.idl b/uriloader/exthandler/nsIHelperAppLauncherDialog.idl
index f8190e7..d15f133 100644
--- a/uriloader/exthandler/nsIHelperAppLauncherDialog.idl
+++ b/uriloader/exthandler/nsIHelperAppLauncherDialog.idl
@@ -21,7 +21,7 @@ interface nsIFile;
  * will access methods of the nsIHelperAppLauncher passed in to show()
  * in order to cause a "save to disk" or "open using" action.
  */
-[scriptable, uuid(bfc739f3-8d75-4034-a6f8-1039a5996bad)]
+[scriptable, uuid(3ae4dca8-ac91-4891-adcf-3fbebed6170e)]
 interface nsIHelperAppLauncherDialog : nsISupports {
   /**
    * This request is passed to the helper app dialog because Gecko can not
@@ -58,6 +58,32 @@ interface nsIHelperAppLauncherDialog : nsISupports {
             in unsigned long aReason);
 
   /**
+   * Invoke a save-to-file dialog instead of the full fledged helper app dialog.
+   * Returns the a nsIFile for the file name/location selected.
+   *
+   * @param aLauncher
+   *        A nsIHelperAppLauncher to be invoked when a file is selected.
+   * @param aWindowContext
+   *        Window associated with action.
+   * @param aDefaultFileName
+   *        Default file name to provide (can be null)
+   * @param aSuggestedFileExtension
+   *        Sugested file extension
+   * @param aForcePrompt
+   *        Set to true to force prompting the user for thet file
+   *        name/location, otherwise perferences may control if the user is
+   *        prompted.
+   *
+   * @throws NS_ERROR_NOT_AVAILABLE if the async version of this function
+   *                                should be used.
+   */
+  nsIFile promptForSaveToFile(in nsIHelperAppLauncher aLauncher,
+                              in nsISupports aWindowContext,
+                              in wstring aDefaultFileName,
+                              in wstring aSuggestedFileExtension,
+                              in boolean aForcePrompt);
+
+  /**
    * Async invoke a save-to-file dialog instead of the full fledged helper app
    * dialog. When the file is chosen (or the dialog is closed), the callback
    * in aLauncher (aLauncher.saveDestinationAvailable) is called with the
diff --git a/uriloader/exthandler/tests/mochitest/test_unsafeBidiChars.xhtml b/uriloader/exthandler/tests/mochitest/test_unsafeBidiChars.xhtml
index 3e5fa41..6363e55 100644
--- a/uriloader/exthandler/tests/mochitest/test_unsafeBidiChars.xhtml
+++ b/uriloader/exthandler/tests/mochitest/test_unsafeBidiChars.xhtml
@@ -84,6 +84,9 @@ function load() {
          "The filename should be correctly sanitized");
       gCallback();
     },
+    promptForSaveToFile: function(aLauncher, aWindowContext, aDefaultFileName, aSuggestedFileExtension, aForcePrompt) {
+      return null;
+    },
     QueryInterface: function(aIID) {
       netscape.security.PrivilegeManager.enablePrivilege('UniversalXPConnect');
       if (aIID.equals(SpecialPowers.Ci.nsISupports) ||
diff --git a/uriloader/exthandler/tests/unit_ipc/test_encoding.js b/uriloader/exthandler/tests/unit_ipc/test_encoding.js
index 35dcf5a..a4cad91 100644
--- a/uriloader/exthandler/tests/unit_ipc/test_encoding.js
+++ b/uriloader/exthandler/tests/unit_ipc/test_encoding.js
@@ -38,7 +38,9 @@ HelperAppDlg.prototype = {
   show: function (launcher, ctx, reason, usePrivateUI) {
     launcher.MIMEInfo.preferredAction = Ci.nsIMIMEInfo.saveToFile;
     launcher.launchWithApplication(null, false);
-  }
+  },
+
+  promptForSaveToFile: function (launcher, ctx, defaultFile, suggestedExtension, forcePrompt) { }
 }
 
 // Override the download-manager-ui to prevent anyone from trying to open
-- 
2.7.4

